### ISSUE TITLE
Handle relationships to terms in other ontologies that are not imported

### ISSUE THREAD

**Issue by mpreusse** [althonos/pronto#89]
Some OBO files on OboFoundry have relationships to `Terms` from other ontologies. Unfortunately, some of those OBO files do not properly import the target ontology (e.g. the `basic.obo` of UBERON, http://www.obofoundry.org/ontology/uberon.html).

As far as I understand this is an issue of the ontology and they should fix the OBO file.

The problem when working with these files in `pronto` is that you cannot access the `.relationships` property of a `Term` if *at least one* relationship has an unidentified target. The `.relationships` property tries to get an instance of each target `Term` from the ontology. Obviously, this is not possible if the target is not part of the ontology and not imported. Hence, accessing `.relationships` escalates the `KeyError` from `Ontology.get_term()`.

It is possible to catch this `KeyError` but you have to wrap the try/except around the `.relationships` call. You cannot access the other relationships because one relationship target cannot be instantiated:

```python
try:
    for reltype, target in some_term.relationships.items():
        # do something
except KeyError as e:
    # handle error
```

If a `Term` has one relationship to an unidentified target and multiple correct relationships it would be great to catch the error on the level of the individual relationship.

Something like:

```python
for rel in some_term.relationships:
    try:
        target = rel.get_target()
    except KeyError as e:
        # this rel does not work but we can try the next one
```

Maybe this is already possible with pronto but I did ot find a way to handle this.

**Comment by althonos** [comment]
Hi @mpreusse ,

this is indeed an error on their side, normally all terms should be either reachable from an import, or declared in the source file.

Right now, an Ontology is stored on two layers: a data layer that has the raw (textual) information extracted from the source file, and a view-layer that let's you see referenced information and edit it without having to be concerned with the data layer. For now, I declared a `Term.relationship` to simply yield a `Mapping[Relationship, FrozenSet["Term"]]`, but it definitely needs its own data structure to allow editing a `Term` relationship, and so on, as discussed in #84. It's just that I didn't have time to implement it.

**Comment by mpreusse** [comment]
Thanks for the update. I started looking into the code, maybe I can try to contribute something. Do you have a sort of high-level dev intro? I.e. basic design principles, running tests etc.

**Comment by althonos** [comment]
Hi @mpreusse 

Sorry for the delay but I had a lot of stuff to do work-wise in these recent months. I finally got around writing the big refactor I wanted for the next version of `pronto`, and I implemented a custom collection to store the `relationships` of an `Entity` rather than using an `immutabledict`.

Now that the creation of the `TermSet` representing targets is delegated to the `__getitem__` call, you will be able to do as you proposed earlier:
```python
for rel in some_term.relationships:
    try:
        targets = some_term.relationships[rel]
    except KeyError as e:
        pass
        # this rel does not work but we can try the next one
````

This will be available in the `v2.3.0` release.

**Comment by mpreusse** [comment]
Thank you for the refactor and your fantastic work on pronto!
